Approved/implicitly-scoped-qubit-allocation.md

```diff
@@ -0,0 +1,342 @@
+        }
+    }
+}
+```
+
+Example 5a: A qubit allocated in a `using` block but released before the end of the outer block.
+
+```qsharp
+operation DifferentLifetime() : Result {
+    mutable r = Zero;
+    using (q = Qubit()) {
+        // Apply operations...
+        set r = M(q);
+    }
+    // Classical post-processing...
+    return r;
+}
+```
+
+## Proposed Modification
+
+The new `use` and `borrow` statements behave identically to `using` and `borrowing`, but the new block is implicitly defined to start at the statement and end at the end of the current block:
+
+```qsharp
+operation FlipCoin() : Result {
+    let x = 10;
+    use q = Qubit();
+    H(q);
+    return M(q);
+}
+```
+
+is equivalent to
+
+```qsharp
+operation FlipCoin() : Unit {
+    let x = 10;
+    using (q = Qubit()) {
+        H(q);
+        return M(q);
+    }
+}
+```
+
+In combination with the new scope statement, the lifetime control provided by `using` can be obtained:
+
+```qsharp
+operation FlipCoin() : Result {
+    mutable r = Zero;
+    {
+        use q = Qubit();
+        H(q);
+        set r = M(q);
+    }
+    return r;
+}
+```
+
+is equivalent to
+
+```qsharp
+operation FlipCoin() : Result {
+    mutable r = Zero;
+    if (true) {
+        using (q = Qubit()) {
+            H(q);
+            set r = M(q);
+        }
+    }
+    return r;
+}
+```
+
+or more simply
+
+```qsharp
+operation FlipCoin() : Result {
+    mutable r = Zero;
+    using (q = Qubit()) {
+        H(q);
+        set r = M(q);
+    }
+    return r;
+}
+```
+
+Each example below is equivalent to the corresponding example in the [Current Status](#current-status) section with the same number.
+
+### Examples
+
+Example 1b: A single qubit allocated and released at the same time as the end of the outer block.
+
+```qsharp
+operation FlipCoin() : Result {
+    use q = Qubit();
+    H(q);
+    return M(q); // q is released here.
+}
+```
+
+Example 2b: Many qubits allocated in a single `use` statement with tuple destructuring.
+
+```qsharp
+operation QubitTuple(n : Int) : Result {
+    use (a, b, c) = (Qubit(), Qubit[2 * n + 1], Qubit[n]);
+    // ...
+    return M(a); // a, b, and c are released here.
+}
+```
+
+Example 3b: Many qubits allocated with a separate `use` statement for each variable.
+
+```qsharp
+operation NestedBlocks(n : Int) : Result {
+    use a = Qubit();
+    use b = Qubit[2 * n + 1];
+    use c = Qubit[n];
+    // ...
+    return M(q); // a, b, and c are released here.
+}
+```
+
+Example 5b: A clean qubit and a borrowed qubit. This can only be expressed with multiple statements, not with tuple destructuring.
+
+```qsharp
+operation UsingAndBorrowing() : Result {
+    use a = Qubit();
+    borrow b = Qubit();
+    // ...
+    return M(a); // a and b are released here.
+}
+```
+
+Example 5b: A qubit allocated in new block and released before the end of the outer block.
+
+```qsharp
+operation DifferentLifetime() : Result {
+    mutable r = Zero;
+    {
+        use q = Qubit();
```

Comment:
The one advantage I see to the previous approach over this example is that it's obvious when looking at the start of the block that the block is intended to delineate a qubit lifetime. Though it would be bad style to do so, I could imagine someone writing code like the following, in which case it would be harder to identify the qubit lifetime:

```Q#
mutable r = Zero;
{
     let foo = f(r);
     set r = bar(foo);
     use q = Qubit();
}
```
In that case, it's not strictly _wrong_ to put classical precomputation inside the nested `{}`, but it does make it harder to read what the purpose of the new scope is.

Reply:
I actually wouldn't consider this bad style, necessarily - assuming `q` is actually used for something after it's allocated. :) Perhaps you want to do something with `foo` but limit its use to that scope. I think it's OK to use the scope statement for scoping of both classical and quantum variables.

Reply:
The scope statement has been removed from the proposal, so this syntax is no longer valid.